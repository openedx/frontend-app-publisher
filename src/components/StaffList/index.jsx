import React from 'react';
import PropTypes from 'prop-types';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';

import { Staffer, getStafferName } from '../Staffer';


class StaffList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      staffList: props.input.value || [], // requires staffList to be an array
    };

    this.handleRemove = this.handleRemove.bind(this);
    this.onDragEnd = this.onDragEnd.bind(this);
  }

  onDragEnd(result) {
    // Element was dropped outside the list.
    if (!result.destination) {
      return;
    }
    // Element has not actually moved.
    if (result.destination.index === result.source.index) {
      return;
    }
    const newStaffList = Array.from(this.state.staffList);
    const [removed] = newStaffList.splice(result.source.index, 1);
    newStaffList.splice(result.destination.index, 0, removed);
    this.setState({
      staffList: newStaffList,
    }, () => this.props.input.onChange(this.state.staffList));
  }

  handleRemove(uuid) {
    this.setState({
      staffList: this.state.staffList.filter(staffer => staffer.uuid !== uuid),
    }, () => this.props.input.onChange(this.state.staffList));
  }

  render() {
    const {
      staffList,
    } = this.state;

    return (
      <React.Fragment>
        Staff
        <DragDropContext onDragEnd={this.onDragEnd}>
          <Droppable droppableId="StaffList" direction="vertical">
            {provided => (
              <div className="staff-list container" ref={provided.innerRef} {...provided.droppableProps}>
                {staffList && staffList.map((staffer, index) => (
                  <Draggable draggableId={getStafferName(staffer)} index={index} key={staffer.uuid}>
                    {draggableProvided => (
                      <div
                        className="staffer-wrapper col-12 my-2"
                        ref={draggableProvided.innerRef}
                        {...draggableProvided.dragHandleProps}
                        {...draggableProvided.draggableProps}
                      >
                        <Staffer
                          onRemove={this.handleRemove}
                          staffer={staffer}
                        />
                      </div>
                    )}
                  </Draggable>
                ))}
                {provided.placeholder}
              </div>
            )}
          </Droppable>
        </DragDropContext>
      </React.Fragment>
    );
  }
}

StaffList.propTypes = {
  input: PropTypes.shape({
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(PropTypes.shape({
        uuid: PropTypes.string.isRequired,
        given_name: PropTypes.string.isRequired,
        family_name: PropTypes.string,
        profile_image_url: PropTypes.string.isRequired,
      })),
    ]).isRequired,
    onChange: PropTypes.func.isRequired,
  }).isRequired,
};

export default StaffList;
