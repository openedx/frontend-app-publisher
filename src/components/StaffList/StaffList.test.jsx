import React from 'react';
import { mount, shallow } from 'enzyme';

import StaffList from './index';


const input = {
  value: [
    {
      uuid: '6f23f2f8-10dd-454a-8497-2ba972c980c4',
      given_name: 'First',
      family_name: 'Last',
      profile_image_url: '/media/people/profile_images/6f23f2f8-10dd-454a-8497-2ba972c980c4-a411afec9477.jpeg',
    },
    {
      uuid: '17d0e2c0-9a02-421b-93bf-d081339090cc',
      given_name: 'Pippa',
      family_name: null,
      profile_image_url: '/media/people/profile_images/17d0e2c0-9a02-421b-93bf-d081339090cc-68912d27b6e7.jpeg',
    },
    {
      uuid: '2aba6189-ad7e-45a8-b269-bea071b80391',
      given_name: 'Dave',
      family_name: 'Grohl',
      profile_image_url: '/media/people/profile_images/2aba6189-ad7e-45a8-b269-bea071b80391-11df6812f839.png',
    },
  ],
  onChange: jest.fn(),
};

describe('StaffList', () => {
  afterEach(() => {
    // Clear onChange's call count after each test
    input.onChange.mockClear();
  });

  it('renders a grid of staff members', () => {
    const component = shallow(<StaffList input={input} />);
    expect(component).toMatchSnapshot();
  });

  it('correctly handles removing members of the staff', () => {
    const component = mount(<StaffList input={input} />);
    let staffers = component.find('Staffer');
    expect(staffers).toHaveLength(input.value.length);

    // Delete the first staffer
    const firstStaffer = staffers.at(0);
    firstStaffer.find('.js-delete-btn').simulate('click');

    // Verify that the onChange method has been called
    expect(input.onChange).toBeCalled();

    // Verify that the first staffer has been removed
    staffers = component.find('Staffer');
    expect(staffers).toHaveLength(input.value.length - 1);
    const newFirstStaffer = staffers.at(0);
    expect(newFirstStaffer).not.toEqual(firstStaffer);
  });

  it('correctly handles reordering members of the staff', () => {
    const component = mount(<StaffList input={input} />);
    // Find the first staffer.
    const firstStaffer = component.state().staffList[0].uuid;

    const result = {
      source: {
        index: 0,
      },
      destination: {
        index: 2,
      },
    };
    // Pretend we dragged the first staffer to the end.
    component.instance().onDragEnd(result);

    // Verify that the onChange method has been called
    expect(input.onChange).toBeCalled();

    // Verify that it is on the end.
    expect(firstStaffer).toEqual(component.state().staffList[2].uuid);
  });

  it('does not re-order when dragged outside of the list', () => {
    const component = mount(<StaffList input={input} />);
    // Find the first staffer.
    const firstStaffer = component.state().staffList[0].uuid;

    const result = {
      source: {
        index: 0,
      },
    };
    // Pretend we dragged the first staffer outside the list.
    component.instance().onDragEnd(result);

    // Verify that the onChange method has NOT been called
    expect(input.onChange).not.toBeCalled();
    expect(firstStaffer).toEqual(component.state().staffList[0].uuid);
  });

  it('does not re-order when dragged to the same position', () => {
    const component = mount(<StaffList input={input} />);
    // Find the first staffer.
    const firstStaffer = component.state().staffList[0].uuid;

    const result = {
      source: {
        index: 0,
      },
      destination: {
        index: 0,
      },
    };
    // Pretend we dragged the first staffer to their original position.
    component.instance().onDragEnd(result);
    // Verify that the onChange method has NOT been called
    expect(input.onChange).not.toBeCalled();
    expect(firstStaffer).toEqual(component.state().staffList[0].uuid);
  });
});
