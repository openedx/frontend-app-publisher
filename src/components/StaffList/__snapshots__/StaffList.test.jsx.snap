// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaffList renders a grid of staff members 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StaffList
    input={
      Object {
        "onChange": [MockFunction],
        "value": Array [
          Object {
            "family_name": "Last",
            "given_name": "First",
            "profile_image_url": "/media/people/profile_images/6f23f2f8-10dd-454a-8497-2ba972c980c4-a411afec9477.jpeg",
            "uuid": "6f23f2f8-10dd-454a-8497-2ba972c980c4",
          },
          Object {
            "family_name": null,
            "given_name": "Pippa",
            "profile_image_url": "/media/people/profile_images/17d0e2c0-9a02-421b-93bf-d081339090cc-68912d27b6e7.jpeg",
            "uuid": "17d0e2c0-9a02-421b-93bf-d081339090cc",
          },
          Object {
            "family_name": "Grohl",
            "given_name": "Dave",
            "profile_image_url": "/media/people/profile_images/2aba6189-ad7e-45a8-b269-bea071b80391-11df6812f839.png",
            "uuid": "2aba6189-ad7e-45a8-b269-bea071b80391",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        "Staff",
        <DragDropContext
          onDragEnd={[Function]}
        >
          <Connect(Droppable)
            direction="vertical"
            droppableId="StaffList"
            ignoreContainerClipping={false}
            isCombineEnabled={false}
            isDropDisabled={false}
            type="DEFAULT"
          >
            [Function]
          </Connect(Droppable)>
        </DragDropContext>,
      ],
    },
    "ref": null,
    "rendered": Array [
      "Staff",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Connect(Droppable)
            direction="vertical"
            droppableId="StaffList"
            ignoreContainerClipping={false}
            isCombineEnabled={false}
            isDropDisabled={false}
            type="DEFAULT"
          >
            [Function]
          </Connect(Droppable)>,
          "onDragEnd": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": [Function],
            "direction": "vertical",
            "droppableId": "StaffList",
            "ignoreContainerClipping": false,
            "isCombineEnabled": false,
            "isDropDisabled": false,
            "type": "DEFAULT",
          },
          "ref": null,
          "rendered": [Function],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          "Staff",
          <DragDropContext
            onDragEnd={[Function]}
          >
            <Connect(Droppable)
              direction="vertical"
              droppableId="StaffList"
              ignoreContainerClipping={false}
              isCombineEnabled={false}
              isDropDisabled={false}
              type="DEFAULT"
            >
              [Function]
            </Connect(Droppable)>
          </DragDropContext>,
        ],
      },
      "ref": null,
      "rendered": Array [
        "Staff",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Connect(Droppable)
              direction="vertical"
              droppableId="StaffList"
              ignoreContainerClipping={false}
              isCombineEnabled={false}
              isDropDisabled={false}
              type="DEFAULT"
            >
              [Function]
            </Connect(Droppable)>,
            "onDragEnd": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": [Function],
              "direction": "vertical",
              "droppableId": "StaffList",
              "ignoreContainerClipping": false,
              "isCombineEnabled": false,
              "isDropDisabled": false,
              "type": "DEFAULT",
            },
            "ref": null,
            "rendered": [Function],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
